File: packages/types/src/service-desk.ts
/**
 * Copyright (c) 2023-present Plane Software, Inc. and contributors
 * SPDX-License-Identifier: AGPL-3.0-only
 * See the LICENSE file for details.
 */

export type TServiceDeskNotifyMode = "NONE" | "ADMINS" | "MEMBERS" | "CUSTOM";

export type TServiceDeskConfig = {
  id: string;
  workspace: string;
  project: string;
  mailbox_email: string;
  is_enabled: boolean;
  notify_mode: TServiceDeskNotifyMode;
  notify_user_ids: string[];
  last_synced_at: string | null;
  created_at: string;
  updated_at: string;
};

export type TEmailDirection = "INBOUND" | "OUTBOUND";

export type TEmailDeliveryStatus = "RECEIVED" | "PENDING" | "SENT" | "FAILED";

export type TIssueEmailMessage = {
  id: string;
  thread: string;
  direction: TEmailDirection;
  status: TEmailDeliveryStatus;
  from_email: string;
  to_emails: string[];
  cc_emails: string[];
  subject: string;
  body_text: string;
  comment: string | null;
  error: string | null;
  created_at: string;
};

export type TIssueEmailThread = {
  id: string;
  workspace: string;
  project: string;
  issue: string;
  mailbox_email: string;
  creator_email: string;
  creator_name: string;
  subject: string;
  to_emails: string[];
  cc_emails: string[];
  messages: TIssueEmailMessage[];
  created_at: string;
  updated_at: string;
};

File: apps/api/plane/app/serializers/service_desk.py
# Copyright (c) 2023-present Plane Software, Inc. and contributors
# SPDX-License-Identifier: AGPL-3.0-only
# See the LICENSE file for details.

# Module imports
from plane.db.models import IssueEmailMessage, IssueEmailThread, ServiceDeskConfig

from .base import BaseSerializer


class ServiceDeskConfigSerializer(BaseSerializer):
    class Meta:
        model = ServiceDeskConfig
        fields = [
            "id",
            "workspace",
            "project",
            "mailbox_email",
            "is_enabled",
            "notify_mode",
            "notify_user_ids",
            "last_synced_at",
            "created_at",
            "updated_at",
        ]
        read_only_fields = ["id", "workspace", "project", "last_synced_at", "created_at", "updated_at"]


class IssueEmailMessageSerializer(BaseSerializer):
    class Meta:
        model = IssueEmailMessage
        fields = [
            "id",
            "thread",
            "direction",
            "status",
            "from_email",
            "to_emails",
            "cc_emails",
            "subject",
            "body_text",
            "comment",
            "attachments",
            "error",
            "created_at",
        ]
        read_only_fields = fields


class IssueEmailThreadSerializer(BaseSerializer):
    messages = IssueEmailMessageSerializer(many=True, read_only=True)

    class Meta:
        model = IssueEmailThread
        fields = [
            "id",
            "workspace",
            "project",
            "issue",
            "mailbox_email",
            "creator_email",
            "creator_name",
            "subject",
            "to_emails",
            "cc_emails",
            "messages",
            "created_at",
            "updated_at",
        ]
        read_only_fields = [
            "id",
            "workspace",
            "project",
            "issue",
            "mailbox_email",
            "creator_email",
            "creator_name",
            "subject",
            "messages",
            "created_at",
            "updated_at",
        ]
